### DISCUSSION TITLE
DMR TIER 1/2

### DISCUSSION THREAD

**Discussion by alwinhb** [qradiolink/dmrtc#2] [Q&A] [answered]
I could see that this repository is precisely for TIER 3. I need to check for Tier 1 and Tier 2 level implementation @kantooon

**Comment by kantooon** [comment, marked as answer]
Well, sure, you can check, but I think the README is pretty clear on the purpose of this code.

**Comment by alwinhb** [comment]
@kantooon Thanks `Adrian` for the help can you share some references for Tier 2 DMR source code(s)?

That can be executed.

**Reply by kantooon** [reply]
MMDVMHost and MMDVM by G4KLX implement Tier 2.
https://github.com/g4klx/

**Reply by alwinhb** [reply]
@kantooon thanks for the quick response, But MMDVM is quite wide I am only looking for `DMR ( Tier 2 )` Specific source code.

**Reply by kantooon** [reply]
Well, unfortunately MMDVM is the only open source implementation I know about.

**Comment by alwinhb** [comment]
@kantooon Any references for 32 bit crc for DMR Protocol

**Reply by kantooon** [reply]
Yes, see the files crc32.cpp and crc32.h in src/

**Reply by alwinhb** [reply]
Cool i see two functions primarily, crc32_update and finalize.

Update is to add 32 bit crc and finalize is to check/validate? Please guide.

**Reply by kantooon** [reply]
Yeah, something like that. Share your code, then we'll see how it needs to be used there. Do you have a Github repo link?

**Reply by alwinhb** [reply]
It isn't there on my Git, I can use the existing code, If you can confirm the usage of functions

`crc32_update`  - To add 32 bit CRC
`finalize`   - To validate the CRC created with crc32_update function.

**Reply by kantooon** [reply]
Ok, I understand. That's not a problem, but I can't provide free support for a non open-source project. You can hire me as a consultant though, contact me privately if you want to discuss fees and contract details.

**Reply by alwinhb** [reply]
Can you help me with your email to discuss further?

I just need understanding on the usage of these two functions - `crc32_update`  and `finalize`

**Reply by kantooon** [reply]
Ok, fine, to me it looks like you're in over your head if you have to ask that, but here goes. 

You're supposed to call `crc_init()` first to initialize the value, but here that won't  actually do anything since there is no initial remainder. `crc_update()` must be called with blocks of data of known length, maybe in a loop if the data is split into multiple blocks. FInally, `call crc_finalize()` to compute the final value, including the DMR specific byte rearanging. You can find more details in annex `B.3.9 32-bit CRC calculation` of TS 102 361-1.

**Reply by alwinhb** [reply]
@kantooon Thanks for the explanation! Can you help me with your email address as well, Good to hear that you can help as a consultant/support for the tech projects.

**Comment by alwinhb** [comment]
@kantooon Can you guide me with Golay 8-20. Algorithm. If I provide the input to file. I am not getting 20 bits output as expected. May be I am missing something on it.
